hubspot_sync/tasks.py

```diff
@@ -0,0 +1,181 @@
+"""
+Hubspot tasks
+"""
+import logging
+import re
+import time
+
+from math import ceil
+from typing import List, Tuple
+
+import celery
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.contrib.contenttypes.models import ContentType
+from hubspot.crm.objects import BatchInputSimplePublicObjectInput
+from mitol.common.utils.collections import chunks
+from mitol.hubspot_api.api import HubspotApi
+from mitol.hubspot_api.models import HubspotObject
+
+from applications.models import BootcampApplication
+from main.celery import app
+
+from hubspot_sync import api
+
+log = logging.getLogger(__name__)
+
+HUBSPOT_SYNC_URL = "/extensions/ecomm/v1/sync-messages"
+ASSOCIATED_DEAL_RE = re.compile(r"\[hs_assoc__deal_id: (.+)\]")
+
+
+def max_concurrent_chunk_size(obj_count: int) -> int:
+    """Divide number of objects by max concurrent tasks for chunk size"""
+    return ceil(obj_count / settings.HUBSPOT_MAX_CONCURRENT_TASKS)
+
+
+def batched_chunks(
+    hubspot_type: str, batch_ids: List[int or (int, str)]
+) -> List[List[int or str]]:
+    """If list of ids exceed max allowed in a batch API call, chunk them up"""
+    max_chunk_size = 10 if hubspot_type == api.HubspotObjectType.CONTACTS.value else 100
+    if len(batch_ids) <= max_chunk_size:
+        return [batch_ids]
+    return chunks(batch_ids, chunk_size=max_chunk_size)
+
+
+@app.task
+def sync_contact_with_hubspot(user_id):
+    """Send a sync-message to sync a user with a hubspot_sync contact"""
+    return api.sync_contact_with_hubspot(user_id).id
+
+
+@app.task
+def sync_product_with_hubspot(bootcamp_run_id):
+    """Send a sync-message to sync a BootcampRun with a hubspot_sync product"""
+    return api.sync_product_with_hubspot(bootcamp_run_id).id
+
+
+@app.task
+def sync_deal_with_hubspot(application_id):
+    """Sync the application to a hubspot_sync deal and line"""
+    return api.sync_deal_with_hubspot(application_id).id
+
+
+@app.task
+def batch_upsert_hubspot_deals_chunked(ids: List[int]):
+    """Batch sync hubspot objects with matching object ids"""
+    results = []
+    for application in BootcampApplication.objects.filter(id__in=ids):
+        results.append(api.sync_deal_with_hubspot(application.id).id)
+        time.sleep(settings.HUBSPOT_TASK_DELAY / 1000)
+    return results
+
+
+@app.task(bind=True)
+def batch_upsert_hubspot_deals(self, create: bool):
+    """Batch create/update deals in hubspot"""
+    content_type = ContentType.objects.get_for_model(BootcampApplication)
+    synced_ids = HubspotObject.objects.filter(content_type=content_type).values_list(
+        "object_id", flat=True
+    )
+    unsynced_ids = BootcampApplication.objects.exclude(id__in=synced_ids).values_list(
+        "id", flat=True
+    )
+    object_ids = sorted(unsynced_ids if create else synced_ids)
+    # Try to avoid too many consecutive tasks that could trigger rate limiting
+    chunk_size = max_concurrent_chunk_size(len(object_ids))
```

Comment:
Does hubspot have an upper limit on their batch api for how many objects they'll accept in a given api request?

What do you think about any/all of the following changes?

- Use a fixed batch size (ideally if they document this, it'd be whatever the API limit is)
- Use celery's `rate_limit` but expose it as a configurable environment variable
- I think we probably still need to catch HTTP 429 errors, which indicate we've hit the rate limit despite our best efforts, and retry the task on a delay, ideally with a backoff which appears to be supported natively now: https://docs.celeryq.dev/en/stable/userguide/tasks.html?highlight=backoff#Task.retry_backoff

Reply:
10 for contacts, 100 for everything else.  

That is handled [in this function](https://github.com/mitodl/bootcamp-ecommerce/pull/1374/commits/147c84183b7904d2fc65a36c1a7e8de8e6ec06d1#diff-69ccb97d890c67f3920dcc34d82c30380486d07832dc5f12b0cf9bafbda7e13cR101), which is only called for contacts and products.  I decided to handle deals (aka Orders and B2BOrders) differently because of the associations that also need to be set up for them (contact-> deal, lines->deal) via separate api calls.  

The hubspot api client has a [builtin retry function ](https://github.com/mitodl/ol-django/blob/957429b024e1f5af3facc2c38b1153a3361bb965/src/mitol/hubspot_api/api.py#L79-L81)